### DISCUSSION TITLE
MacOS Support

### DISCUSSION THREAD

**Discussion by roaldarbol** [bonsai-rx/bonsai#1436] [General]
Hi! 
A colleague and I are in the early stages of planning a workshop on animal tracking, and we are considering including Bonsai. However, one of our criteria is that the software we introduce should be able to run cross-platform, with installation being relatively painfree. So it has prompted [a question](https://groups.google.com/g/bonsai-users/c/5E6KQPzvsRs) to resurface: 

- Is it on the roadmap to get Bonsai supported on **MacOS**? 
- And if so, does that include the editor?

Of course, I'm hoping the answer to both is yes, but I know there are many other things to prioritise. Getting the editor to work would be an immense step! One option I noticed going through the old threads and issues is that it should(?) be possible to run it in a Virtual Machine, but I can't find that documented anywhere (see the last point here). 

### Updates for branch `3.0`
It seems that the `3.0` branch of Bonsai is far far behind the HEAD of the project, which make me wonder if it is still on the roadmap? One key thing is that as .NET 5.0 has reached end of life status and will no longer be supported, I guess it makes sense to change over to .NET 7.0? I built Bonsai with .NET 7.0 (installed with `brew`), following the steps from the [Google Group conversation](https://groups.google.com/g/bonsai-users/c/5E6KQPzvsRs), just replacing `net5.0` with `net7.0`, and get no compiling errors (although a few warnings - can share if it helps). I can't, however get the examples to run, I guess because `net5.0` is hardcoded in somehow. But I think it will be a relatively simple upgrade (but again, I'm not an expert in .NET, so take that with a grain of salt).

### Documentation
Additionally, since it seems (from what I can gather from [issues] and [conversations](), it is easy to run on Linux, it would be a great thing to include in the [installation documentation](https://bonsai-rx.org/docs/articles/installation.html). It was also mentioned that Bonsai can run in Virtual Machines, but that doesn't seem to be documented (please correct me if I'm wrong). That would be really great to have documented, maybe with a short guide - please let me know if I can help out with these points, as I will be documenting stuff anyways for the workshop. 

Looking forward to hearing back and continuing the discussion. 😊
Cheers,
Mikkel

**Comment by glopesdev** [comment]
@roaldarbol Indeed, both are still on the roadmap, but there have been a few tweaks which actually is making the transition smoother rather than a fully separate development.

The branch 3.0 is indeed stale right now (this will change) but the main reason is that the integration to the cross-platform .NET has already started in the main branch! You might have noticed that some of the core NuGet packages (e.g. Core and System) are already targeting both `net472` and `net6.0`.

This means that with some work you could actually already run a simple Bonsai workflow on a MacOS system by installing .NET 6.0 and linking against those cross-platform binaries. These don't include the editor yet, and to be fair they won't be incredibly useful until we expand support to the remaining packages which might take a while. However there is at least a clear path forward on what needs to be done that we want to keep pursuing, even for the Windows version, since it would bring significant performance improvements and other benefits.

We aim to be following the stable releases of .NET, so we will be updating this target with the upcoming .NET 8.0 when it releases later this year. This release will include a revamped cross-platform Rx.NET which we are really excited about as it finally solves some important pain points which were making our transition more difficult. 

Re. Linux we just published updated [Linux installation steps](https://github.com/orgs/bonsai-rx/discussions/1101) which we do want to eventually move to the official documentation, especially since for the latest version 2.8 we did do a lot of work to significantly increase compatibility with the Mono framework on both AMD64 and ARM64 architectures, and these include the editor.

As for VMs, this scenario I have to say I haven't tested extensively so I cannot really say how well it works, but happy to hear about difficulties or experiences others have had with this option and we can document and polish them first in these discussions and later potentially move them to the docs.

Happy to elaborate further on any detail you might be curious about.

**Reply by roaldarbol** [reply]
Really happy to hear all of it! It seems you've made great progress, and I'll be first in line when you need testing "in the wild".😉 May I dare to ask if you have got a vague idea of an ETA for these?

> The branch 3.0 is indeed stale right now (this will change) but the main reason is that the integration to the cross-platform .NET has already started in the main branch! You might have noticed that some of the core NuGet packages (e.g. Core and System) are already targeting both net472 and net6.0.

That's really good news! 

> This release will include a revamped cross-platform Rx.NET which we are really excited about as it finally solves some important pain points which were making our transition more difficult.

Also really good news! 😂

> Re. Linux we just published updated [Linux installation steps](https://github.com/orgs/bonsai-rx/discussions/1101) which we do want to eventually move to the official documentation, especially since for the latest version 2.8 we did do a lot of work to significantly increase compatibility with the Mono framework on both AMD64 and ARM64 architectures, and these include the editor.

Have you had a chance to test if those installation tips would work on Raspberry Pi (either RPi OS or Ubuntu)? I might give that a go again.

> As for VMs, this scenario I have to say I haven't tested extensively so I cannot really say how well it works, but happy to hear about difficulties or experiences others have had with this option and we can document and polish them first in these discussions and later potentially move them to the docs.

I also haven't really played around with it before, and it seems that of course getting Windows running on VirtualBox is a bit trickier than I had hoped. But I will test with Ubuntu and let you know how that goes, maybe next week (have an application and some pilot experiments I need to get sorted first).
